README.rst
======================
Kubernetes Log Watcher
======================

.. image:: https://api.travis-ci.org/zalando-incubator/kubernetes-log-watcher.svg?branch=master
  :target: https://travis-ci.org/zalando-incubator/kubernetes-log-watcher

.. image:: https://codecov.io/gh/zalando-incubator/kubernetes-log-watcher/branch/master/graph/badge.svg
  :target: https://codecov.io/gh/zalando-incubator/kubernetes-log-watcher

Kubernetes Log Watcher is used to facilitate log shipping on a Kubernetes cluster. The watcher will detect containers running on a cluster node, and adjust configurations for number of log configuration agents. ``kubernetes-log-watcher`` is intended to run in a Kubernetes ``DaemonSet`` along with one or more log shippers.

``kubernetes-log-watcher`` comes with builtin configuration agents, however it could be extended to support external agents. Current builtin configuration agents are:

- AppDynamics
- Scalyr
- Symlinker (embeds metadata in symlink filenames, to be extracted by a log shipping agent such as Fluentd)

``kubernetes-log-watcher`` is used in `kubernetes-on-aws <https://github.com/zalando-incubator/kubernetes-on-aws>`_ project.

Components
==========

Watcher
-------

Log watcher is responsible for detecting running containers on a cluster node, extract information from container labels, and supply sufficient information to configuration agents in order to facilitate log shipping. The log watcher could run with one or more configuration agents.

Configuration Agent
-------------------

The agent acts as a plugin, its sole purpose is to adjust log configuration for its log shipper. It is important to note that the configuration agent is not intended to process the logs, it only adjusts configuration files for log shipper (e.g. set log source paths and custom attributes with log events) based on information received from log watcher.

Features & Constraints
======================

* Log watcher accepts one or more log configuration agent.
* Log watcher will skip ``pause`` containers.
* Configuration agents provide the ability to dynamically attach tags/attributes/metadata to logs based on Kubernetes labels.
* **Optionally** follow logs from containers running in pods with a defined list of metadata labels. (optional since 0.14)
* Sync new and stale containers.

Usage
=====

Kubernetes log watcher is intended to run in a ``DaemonSet`` along with one or more log configuration agent (e.g. Scalyr, AppDynamics ...). The DaemonSet will ensure that the watcher is running on every cluster node.


Diagram
-------

The following diagram describes how Log watcher DaemonSet interacts with cluster node containers, Kubernetes API and Log Storage.

.. code-block::

                             +------------------------------------------------------+
    +--------------------+   |                         NODE                         |   +--------------------+
    |       K8s API      |   |  +----------------+              +----------------+  |   |    Log Storage     |
    |                    |   |  |     pod 1      |              |     pod 2      |  |   |                    |
    |                    |   |  |                |              |                |  |   |                    |
    |                    |   |  | +------------+ |              | +------------+ |  |   |                    |
    |                    |   |  | |   Cont 1   | |              | |   Cont 2   | |  |   |                    |
    |                    |   |  | +------+-----+ |              | +--------+---+ |  |   |                    |
    +---------+----------+   |  |        |       |              |          |     |  |   +---------+----------+
              ^              |  |        |       |              |          |     |  |             ^
              |              |  +----------------+              +----------------+  |             |
              |              |           |                                 v        |             |
              |              |           |                                          |             |
              |              |           +-----------[0]--------------> +-----+     |             |
              |              |                                          |LOGS |     |             |
              |              |  +-----------------------------------------------+   |             |
              |              |  | Log watcher pod (DaemonSet)           |     | |   |             |
              |              |  |       +-------[1]----------> +------> +-----+ |   |             |
              |              |  |       |                      |                |   |             |
              |              |  | +-----+-----+          +-----+--------------+ |   |             |
              +-----[2,3]---------+Log|watcher+--[4]---> |     Log shipper    +------------[5]----+
                             |  | +-----------+          +--------------------+ |   |
                             |  +-----------------------------------------------+   |
                             +------------------------------------------------------+

Operation
---------

#. ``Log watcher`` container detects changes in node containers mounted directory (typically ``/var/lib/docker/containers``)
#. ``Log watcher`` container loads containers information with help of Kubernetes API.
#. ``Log watcher`` container will supply containers info to configuration agents.
#. Configuration agents will adjust log configuration for the ``Log shipper`` container (one or more). Log configuration may include containers log file paths and extra attributes/tags attached to the logs.
#. ``Log shippers`` should start following logs from mounted log volume (based on log configuration) and ship logs to the log storage.

Example manifest
----------------

This is an example manifest for shipping logs to Scalyr, with additional Journald monitoring for master processes running on the node.

.. note::

    - This manifest assumes running a Kubernetes cluster version >= 1.5 (as it depends on `initContainer <https://kubernetes.io/docs/concepts/workloads/pods/init-containers/>`_ for initial Scalyr configuration)
    - All shared volumes are of type ``hostPath`` in order to survive pod restarts.
    - Initial Scalyr configuration using ``configMap`` is no longer used as it appears to be reseted to initial values by Kubernetes.

.. code-block:: yaml

    apiVersion: extensions/v1beta1
    kind: DaemonSet
    metadata:
        name: logging-agent
        namespace: kube-system
        labels:
          application: logging-agent
          version: v0.27
          component: logging
    spec:
        selector:
          matchLabels:
            application: logging-agent
        template:
          metadata:
            name: logging-agent
            labels:
              application: logging-agent
              version: v0.27
              component: logging
            annotations:
              scheduler.alpha.kubernetes.io/critical-pod: ''
              scheduler.alpha.kubernetes.io/tolerations: '[{"key":"CriticalAddonsOnly", "operator":"Exists"}]'
              pod.beta.kubernetes.io/init-containers: '[
                {
                  "name": "init-scalyr-config",
                  "image": "busybox",
                  "imagePullPolicy": "IfNotPresent",
                  "command": ["sh", "-c"],
                  "args":
                    - |
                      SCALYR_CONFIG_PATH="/mnt/scalyr/agent.json"
                      if [ -f "$SCALYR_CONFIG_PATH" ]; then
                        echo "Has agent.json with configuration:"
                        cat $SCALYR_CONFIG_PATH;
                      else
                        # Write a minimal configuration which let scalyr agent to start and wait for real configuration
                        echo "Create agent.json with inital configuration:"
                        tee "$SCALYR_CONFIG_PATH" <<EOF
                      {
                          "api_key": "$WATCHER_SCALYR_API_KEY",
                          "scalyr_server": "${WATCHER_SCALYR_SERVER:-https://upload.eu.scalyr.com}",
                          "implicit_agent_process_metrics_monitor": false,
                          "implicit_metric_monitor": false,
                          "monitors": [],
                          "logs": []
                      }
                      EOF
                      # ^^^ "EOF" must be at 0 position after YAML decode
                      fi;

                      SCALYR_CHECKPOINTS_PATH="/mnt/scalyr-agent-checkpoints/checkpoints.json"
                      if [ -f "$SCALYR_CHECKPOINTS_PATH" ]; then
                        echo
                        ls -lah "$SCALYR_CHECKPOINTS_PATH"
                        cat "$SCALYR_CHECKPOINTS_PATH"
                      fi
                  "env":
                    - name: WATCHER_SCALYR_API_KEY
                      value: "<SCALYR-KEY-HERE>"
                  "volumeMounts": [
                    {
                      "name": "scalyr-config",
                      "mountPath": "/mnt/scalyr"
                    },
                    {
                      "name": "scalyr-checkpoint",
                      "mountPath": "/mnt/scalyr-checkpoint"
                    }
                  ]
                }
              ]'
          spec:
            containers:
            - name: log-watcher
              image: registry.opensource.zalan.do/eagleeye/kubernetes-log-watcher:0.27
              env:
              - name: CLUSTER_NODE_NAME
                valueFrom:
                  fieldRef:
                    fieldPath: spec.nodeName

              - name: WATCHER_DEBUG
                value: "true"
              - name: WATCHER_CLUSTER_ID
                value: "kubernetes-cluster-1"

              - name: WATCHER_AGENTS
                value: scalyr
              - name: WATCHER_SCALYR_API_KEY_FILE
                value: "<PATH-TO-SCALYR-KEY-HERE>"
              - name: WATCHER_SCALYR_DEST_PATH
                value: /mnt/scalyr-logs
              - name: WATCHER_SCALYR_CONFIG_PATH
                value: /mnt/scalyr-config/agent.json
              - name: WATCHER_CONFIG
                value: /mnt/scalyr-config/agent.json
              - name: WATCHER_SCALYR_JOURNALD
                value: "true"

              volumeMounts:
              - name: containerlogs
                mountPath: /mnt/containers
                readOnly: true
              - name: scalyr-logs
                mountPath: /mnt/scalyr-logs
                readOnly: false
              - name: scalyr-config
                mountPath: /mnt/scalyr-config

            - name: scalyr-agent

              image: registry.opensource.zalan.do/eagleeye/scalyr-agent:0.2

              volumeMounts:
              - name: containerlogs
                mountPath: /mnt/containers
                readOnly: true
              - name: scalyr-logs
                mountPath: /mnt/scalyr-logs
                readOnly: true
              - name: scalyr-checkpoint
                mountPath: /var/lib/scalyr-agent-2
              - name: scalyr-config
                mountPath: /etc/scalyr-agent-2
                readOnly: true
              - name: journal
                mountPath: /var/log/journal
                readOnly: true

            volumes:
            - name: containerlogs
              hostPath:
                path: /var/lib/docker/containers

            - name: journal
              hostPath:
                path: /var/log/journal

            - name: scalyr-checkpoint
              hostPath:
                path: /var/lib/scalyr-agent

            - name: scalyr-config
              hostPath:
                path: /etc/scalyr-agent

            - name: scalyr-logs
              hostPath:
                path: /var/log/scalyr-agent


Configuration
-------------

Log watcher accepts a set of configuration variables to adjust its behavior. The same applies to builtin configuration agents.

Log watcher
^^^^^^^^^^^

Configuration variables can be set via Env variables:

WATCHER_CONTAINERS_PATH
  Containers directory path mounted from the host (Default: ``/var/lib/docker/containers``)

WATCHER_STRICT_LABELS
  If set then only containers running in pods with the list of metadata labels will be considered for log watching. Value is a comma separated string of label names. (Default is ``''``)

  If no ``application`` label is set then kubernetes-log-watcher will set ``application`` from *pod name*; in order to provide consistent attributes to log configuration agents.

WATCHER_AGENTS
   Comma separated string of required log configuration agents. (Required. Example: "scalyr,appdynamics")

WATCHER_CLUSTER_ID
   Kubernetes Cluster ID.

WATCHER_KUBE_URL
   URL to API proxy service. Service is expected to handle authentication to the Kubernetes cluster. If set, then log-watcher will not use serviceaccount config.

WATCHER_KUBERNETES_UPDATE_CERTIFICATES
   [Deprecated] Call update-ca-certificates for Kubernetes service account ca.crt.

WATCHER_INTERVAL
   Polling interval (secs) for the watcher to detect containers changes. (Default: 60 sec)

WATCHER_DEBUG
   Verbose output. (Default: False)

Scalyr configuration agent
^^^^^^^^^^^^^^^^^^^^^^^^^^

Configuration variables can be set via Env variables:

WATCHER_CONFIG
  Log watcher configuration file (YAML).

WATCHER_SCALYR_API_KEY
  Scalyr API key. (Required).

WATCHER_SCALYR_API_KEY_FILE
  Path to a file with Scalyr API key. (Required).

WATCHER_SCALYR_DEST_PATH
  Scalyr configuration agent will symlink containers logs in this location. This is to provide more friendly name for log files. Typical log file name for a container will be in the form ``<application>-<version>.log``. (Required).

WATCHER_SCALYR_CONFIG_PATH
  Scalyr configuration file path. (Default: ``/etc/scalyr-agent-2/agent.json``)

WATCHER_SCALYR_ENABLE_PROFILING
  If true, the agent will log performance profiling data about itself into a log file.

WATCHER_SCALYR_PARSE_LINES_JSON
  Parse lines lines on the client as JSON. Useful for raw docker logs. (Default: ``False``)

WATCHER_SCALYR_JOURNALD
  Scalyr should follow Journald logs. This is for node system processes log shipping (e.g. docker, kube) (Default: ``False``)

WATCHER_SCALYR_JOURNALD_ATTRIBUTES
  Add attributes to Journald logs. By default ``cluster`` and ``node`` will be added by the configuration agent.

WATCHER_SCALYR_JOURNALD_EXTRA_FIELDS
  Add extra Systemd Journald fields. Should be a JSON string. Example: '{"_COMM": "command"}'

WATCHER_SCALYR_JOURNALD_PATH
  Journald logs path mounted from the host. (Default: ``/var/log/journald``)

WATCHER_SCALYR_JOURNALD_WRITE_RATE
  Journald monitor write rate. (Default: 10000)

WATCHER_SCALYR_JOURNALD_WRITE_BURST
  Journald monitor write burst. (Default: 200000)

Scalyr custom parser
....................

The default parser for container logs is ``json`` parser. In some cases however you might need to assign a `custom Scalyr parser <https://www.scalyr.com/help/config>`_ for specific container. This can be achieved via pod annotations. The following example shows an annotation value that instructs kubernetes-log-watcher to set custom parser ``json-java-parser`` for container ``app-1``.

.. code-block:: yaml

  annotations:
    kubernetes-log-watcher/scalyr-parser: '[{"container": "app-1", "parser": "json-java-parser"}]'

The value of ``kubernetes-log-watcher/scalyr-parser`` annotation should be a json serialized list. If ``container`` value did not match, then default parser is used (i.e. ``json``).

Scalyr sampling rules
....................

Sampling rules enable to only ship a certain pattern that matches a regular expression and specified amount of log percentage to Scalyr. The example shows an expression that matches ``app-1`` and a match expression ``my-expression``. If it's met, only 10% of it will be shipped to Scalyr using a ``sampling_rate`` of ``0.1``.

.. code-block:: yaml

  annotations:
    kubernetes-log-watcher/scalyr-sampling-rules: '[{"container": "app-1", "sampling-rules":[{ "match_expression": "my-expression", "sampling_rate": "0.1" }]}]'


Scalyr log redaction
....................

Redaction rules enable to avoid shipping sensitive data that shouldn't get transferred to Scalyr either getting fully removed from log files or to replace them with specific strings. The first example below shows how matches will be fully removed and the second shows how matches will be replaced with a different string.

.. code-block:: yaml

  annotations:
    kubernetes-log-watcher/scalyr-redaction-rules: '[{"container": "app-1", "redaction-rules":[{ "match_expression": "my-expression" }]}]'
    kubernetes-log-watcher/scalyr-redaction-rules: '[{"container": "app-1", "redaction-rules":[{ "match_expression": "my-expression", "replacement": "replacement-expression" }]}]'

The following redaction rule is added automatically for all containers. It redacts `JSON Web Tokens  <https://tools.ietf.org/html/rfc7519>`_ from all logs.

.. code-block:: json

   {
     "match_expression": "eyJ[a-zA-Z0-9/+_=-]{5,}\\.eyJ[a-zA-Z0-9/+_=-]{5,}\\.[a-zA-Z0-9/+_=-]{5,}",
     "replacement": "+++JWT_TOKEN_REDACTED+++"
   }

AppDynamics configuration agent
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Configuration variables can be set via Env variables:

WATCHER_APPDYNAMICS_DEST_PATH
  AppDynamics job files path. (Required).

AppDynamics configuration agent could also add ``app_name`` and ``tier_name`` if ``appdynamics_app`` and ``appdynamics_tier`` were set in pod metadata labels.

Symlinker configuration agent
^^^^^^^^^^^^^^^^^^^^^^^^^

The Symlinker agent requires only one environment variable:

WATCHER_SYMLINK_DIR
  Base directory where symlink directory structure will be created.


Development
===========

Preferably create a Python 3.5 ``virtualenv``.

.. code-block:: bash

    $ pip install -r requirements.txt
    $ python -m kube_log_watcher --help

Tests
-----

You can use ``pytest``

.. code-block:: bash

    # test requirements
    $ pip install -U flake8 mock pytest pytest_cov codecov>=1.4.0

    $ py.test -v tests/
    $ flake8 .

or via ``tox``

.. code-block:: bash

    $ tox

TODO
====

- Support custom extra/external agents (e.g. ``kube-log-watcher --extra-agent /var/lib/custom-agent.py``)
- Support configuration from config files instead of env variables (e.g. ``kube-log-watcher --config /etc/kube-log-watcher/config.yaml``)
- Support running kube-log-watcher as standalone (release to PyPi)
- Add more configuration agents (logstash, fluentd, etc ...)

All contributions are welcome :)

License
=======

The MIT License (MIT)

Copyright (c) 2016 Zalando SE, https://tech.zalando.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE

CHANGES.txt
================================
Kubernetes-log-watcher changelog
================================

0.42 (2020-02-28)
-----------------

- Load Scalyr API key from file

0.41 (2020-01-29)
-----------------

- Change configuration to support built-in `journald_monitor`

0.38 (2020-01-22)
-----------------

- Add Sentry integration [#109]
- Handle target information stateful and only remove targets if container on disk is removed. [#109]

0.37 (2020-01-10)
-----------------

- Remove application from pod-name fallback [#107]

0.36 (2019-09-04)
-----------------

- Add option to enable profiling [#235]

0.35 (2019-09-04)
-----------------

- Increase max line length to 49900 for scalyr-agent [#235]

0.34 (2019-08-27)
-----------------

- Optimize what are attributes and what are server_attributes [#93]

0.33 (2019-08-22)
-----------------

- Increase compression level to 9 [#92]

0.32 (2019-08-22)
-----------------

- Increase max_allowed_request_size and decrease pipeline_threshold [#91]

0.31 (2019-08-07)
-----------------

- Optimize scalyr-agent sending parameters [#90]

0.30 (2019-07-24)
-----------------

- Rename log-files to include application, component and version [#88]

0.29 (2019-06-19)
-----------------

- revert #85 [#87]

0.28 (2019-06-18)
-----------------

- Set `component` attribute for Scalyr [#84]
- Add scalyr logfiles to scalyr config [#85]

0.27 (2019-05-10)
-----------------

- JSON Web Token redaction from all log files. [#82]


0.26 (2019-03-22)
-----------------

- Add Scalyr server fields cluster_alias and cluster_environment. [#78]


0.25 (2019-02-26)
-----------------

- Clean up dangling symlinks. [#76]


0.24 (2019-02-12)
-----------------

- Add a new Symlinker agent, to be used with Fluentd. [#74]


0.15 (2017-05-30)
-----------------

- Make Journald monitor write rate limit configurable. [#41]


0.14 (2017-05-15)
-----------------

- Drop constraint for application and version labels. [#38]
- More unit test coverage.


0.13 (2017-04-19)
-----------------

- Add support for Scalyr custom parsers using pod annotations. [#36]
- More unit test coverage.
- Deprecate ``update_ca_certificate``.


kube_log_watcher/agents/scalyr.py
"""
Scalyr watcher agent for providing config file and variables required to ship logs to Scalyr.
"""
import binascii
import json
import logging
import os
import shutil

from kube_log_watcher.agents.base import BaseWatcher
from kube_log_watcher.template_loader import load_template

TPL_NAME = 'scalyr.json.jinja2'

SCALYR_CONFIG_PATH = '/etc/scalyr-agent-2/agent.json'

# If exists! we expect serialized json str: '[{"container": "my-container", "parser": "my-custom-parser"}]'
SCALYR_ANNOTATION_PARSER = 'kubernetes-log-watcher/scalyr-parser'
# If exists! we expect serialized json str:
# '[{"container": "my-container", "sampling-rules":[{ "match_expression": "<expression here>",
#  "sampling_rate": "0" }]}]'
SCALYR_ANNOTATION_SAMPLING_RULES = 'kubernetes-log-watcher/scalyr-sampling-rules'
# '[{"container": "my-container", "redaction-rules":[{ "match_expression": "<expression here>" }]}]'
SCALYR_ANNOTATION_REDACTION_RULES = 'kubernetes-log-watcher/scalyr-redaction-rules'
JWT_REDACTION_RULE = {
    "match_expression": "eyJ[a-zA-Z0-9/+_=-]{5,}\\.eyJ[a-zA-Z0-9/+_=-]{5,}\\.[a-zA-Z0-9/+_=-]{5,}",
    "replacement": "+++JWT_TOKEN_REDACTED+++"
}
SCALYR_DEFAULT_PARSER = 'json'
SCALYR_DEFAULT_WRITE_RATE = 10000
SCALYR_DEFAULT_WRITE_BURST = 200000

logger = logging.getLogger('kube_log_watcher')


def container_annotation(annotations, container_name, pod_name, annotation_key, result_key, default=None):
    if annotations and annotation_key in annotations:
        try:
            result_candidates = json.loads(annotations[annotation_key])
            if type(result_candidates) is not list:
                logger.warning(
                    'Scalyr watcher agent found invalid %s annotation in pod: %s. Expected `list` found: `%s`',
                    annotation_key, pod_name, type(result_candidates))
            else:
                for candidate in result_candidates:
                    if candidate.get('container') == container_name:
                        return candidate.get(result_key, default)
        except json.JSONDecodeError:
            logger.exception(
                'Scalyr watcher agent failed to load annotation %s for container %s in pod %s',
                annotation_key, container_name, pod_name)

    return default


def get_parser(annotations, kwargs):
    return container_annotation(annotations=annotations,
                                container_name=kwargs['container_name'],
                                pod_name=kwargs['pod_name'],
                                annotation_key=SCALYR_ANNOTATION_PARSER,
                                result_key='parser',
                                default=SCALYR_DEFAULT_PARSER)


def get_sampling_rules(annotations, kwargs):
    return container_annotation(annotations=annotations,
                                container_name=kwargs['container_name'],
                                pod_name=kwargs['pod_name'],
                                annotation_key=SCALYR_ANNOTATION_SAMPLING_RULES,
                                result_key='sampling-rules',
                                default=None)


def get_redaction_rules(annotations, kwargs):
    rules = container_annotation(annotations=annotations,
                                 container_name=kwargs['container_name'],
                                 pod_name=kwargs['pod_name'],
                                 annotation_key=SCALYR_ANNOTATION_REDACTION_RULES,
                                 result_key='redaction-rules',
                                 default=[])
    if type(rules) is not list:
        logger.warning('Scalyr watcher agent found invalid redaction rule annotation in pod/container: %s/%s. '
                       'Expected `list` found: `%s`', kwargs['pod_name'], kwargs['container_name'], type(rules))
        rules = []
    rules.append(JWT_REDACTION_RULE)
    return rules


class ScalyrAgent(BaseWatcher):
    def __init__(self, configuration):
        cluster_id = configuration['cluster_id']
        self.scalyr_sampling_rules = ScalyrAgent.parse_scalyr_sampling_rules(
            configuration.get('scalyr_sampling_rules') or [],
        )
        self.api_key = os.environ.get('WATCHER_SCALYR_API_KEY')
        self.dest_path = os.environ.get('WATCHER_SCALYR_DEST_PATH')
        self.scalyr_server = os.environ.get('WATCHER_SCALYR_SERVER')
        self.parse_lines_json = os.environ.get('WATCHER_SCALYR_PARSE_LINES_JSON', '').lower() == 'true'
        self.enable_profiling = os.environ.get('WATCHER_SCALYR_ENABLE_PROFILING', '').lower() == 'true'
        cluster_alias = os.environ.get('CLUSTER_ALIAS', 'none')
        cluster_environment = os.environ.get('CLUSTER_ENVIRONMENT', 'production')
        node_name = os.environ.get('CLUSTER_NODE_NAME', 'unknown')

        if not all([self.api_key, self.dest_path]):
            raise RuntimeError('Scalyr watcher agent initialization failed. Env variables WATCHER_SCALYR_API_KEY and '
                               'WATCHER_SCALYR_DEST_PATH must be set.')

        self.config_path = os.environ.get('WATCHER_SCALYR_CONFIG_PATH', SCALYR_CONFIG_PATH)
        if not os.path.isdir(os.path.dirname(self.config_path)):
            raise RuntimeError(
                'Scalyr watcher agent initialization failed. {} config path does not exist.'.format(
                    self.config_path))

        if not os.path.isdir(self.dest_path):
            raise RuntimeError(
                'Scalyr watcher agent initialization failed. {} destination path does not exist.'.format(
                    self.dest_path))
        else:
            watched_containers = os.listdir(self.dest_path)
            logger.info('Scalyr watcher agent found %d watched containers.', len(watched_containers))
            logger.debug('Scalyr watcher agent found the following watched containers: %s', watched_containers)

        self.journald = None
        journald_monitor = os.environ.get('WATCHER_SCALYR_JOURNALD', False)

        if journald_monitor:
            attributes_str = os.environ.get('WATCHER_SCALYR_JOURNALD_ATTRIBUTES', '{}')
            extra_fields_str = os.environ.get('WATCHER_SCALYR_JOURNALD_EXTRA_FIELDS', '{}')
            self.journald = {
                'journal_path': os.environ.get('WATCHER_SCALYR_JOURNALD_PATH'),
                'attributes': json.loads(attributes_str),
                'extra_fields': json.loads(extra_fields_str),
                'write_rate': int(os.environ.get('WATCHER_SCALYR_JOURNALD_WRITE_RATE', SCALYR_DEFAULT_WRITE_RATE)),
                'write_burst': int(os.environ.get('WATCHER_SCALYR_JOURNALD_WRITE_BURST', SCALYR_DEFAULT_WRITE_BURST)),
            }

        self.server_attributes = {
            'serverHost': cluster_id,
            'cluster': cluster_id,
            'cluster_environment': cluster_environment,
            'cluster_alias': cluster_alias,
            'environment': cluster_environment,
            'node': node_name,
            'parser': SCALYR_DEFAULT_PARSER
        }

        self.tpl = load_template(TPL_NAME)
        self.logs = {}
        self._first_run = True

        logger.info('Scalyr watcher agent initialization complete!')

    @staticmethod
    def parse_scalyr_sampling_rules(scalyr_sampling_rules):
        parsed_scalyr_sampling_rules = []

        for scalyr_sampling_rule in scalyr_sampling_rules:
            try:
                if ('probability' in scalyr_sampling_rule) and not (0 <= scalyr_sampling_rule['probability'] <= 1):
                    raise ValueError('`probability` must be between 0 and 1')

                json.loads(scalyr_sampling_rule['value'])
            except (TypeError, KeyError, ValueError) as error:
                logger.warning('Cannot parse rule `%s`: %s', scalyr_sampling_rule, repr(error))
            else:
                parsed_scalyr_sampling_rules.append(scalyr_sampling_rule)

        return parsed_scalyr_sampling_rules

    @property
    def name(self):
        return 'Scalyr'

    @property
    def first_run(self):
        return self._first_run

    def get_scalyr_sampling_rule(self, container_data):
        for scalyr_sampling_rule in self.scalyr_sampling_rules:
            if (
                ('application' in scalyr_sampling_rule)
                and (scalyr_sampling_rule['application'] != container_data['application'])
            ):
                continue

            if (
                ('component' in scalyr_sampling_rule)
                and (scalyr_sampling_rule['component'] != container_data['component'])
            ):
                continue

            if 'probability' in scalyr_sampling_rule:
                container_crc = binascii.crc32(container_data['container_id'].encode())
                if ((container_crc % 100) + 1) > scalyr_sampling_rule['probability'] * 100:
                    continue

            return scalyr_sampling_rule['value']

    def add_log_target(self, target: dict):
        """
        Create our log targets, and pick relevant log fields from ``target['kwargs']``
        """
        log_path = self._adjust_target_log_path(target)
        if not log_path:
            logger.warning('Scalyr watcher agent skipped log config for container(%s) in pod %s.',
                           target['kwargs']['container_name'], target['kwargs']['pod_name'])
            return

        kwargs = target['kwargs']
        annotations = kwargs.get('pod_annotations', {})

        attributes = {
            'application': kwargs['application'],
            'component': kwargs['component'],
            'environment': kwargs['environment'],
            'version': kwargs['version'],
            'release': kwargs['release'],
            'pod': kwargs['pod_name'],
            'namespace': kwargs['namespace'],
            'container': kwargs['container_name'],
            'container_id': kwargs['container_id'],
            'parser': get_parser(annotations, kwargs)
        }

        # Keep only attributes that has value not duplicated in server_attributes
        attributes = {
            k: v
            for k, v in attributes.items()
            if v and (self.server_attributes.get(k) != v)
        }

        sampling_rules = self.get_scalyr_sampling_rule(kwargs)
        if sampling_rules is not None:
            logger.warning('Overwriting container %s (%s/%s) sampling annotation',
                           kwargs['container_id'], kwargs['application'], kwargs['component'])
            annotations[SCALYR_ANNOTATION_SAMPLING_RULES] = sampling_rules

        log = {
            'path': log_path,
            'sampling_rules': get_sampling_rules(annotations, kwargs),
            'redaction_rules': get_redaction_rules(annotations, kwargs),
            'attributes': attributes,
        }

        self.logs[target['id']] = log

    def remove_log_target(self, container_id: str):
        container_dir = os.path.join(self.dest_path, container_id)

        try:
            del self.logs[container_id]
        except KeyError:
            logger.warning('Failed to remove log target: %s', container_id)

        try:
            shutil.rmtree(container_dir)
        except OSError:
            logger.warning('Scalyr watcher agent failed to remove container directory %s', container_dir)

    def flush(self):
        kwargs = {
            'scalyr_key': self.api_key,
            'server_attributes': self.server_attributes,
            'logs': self.logs.values(),
            'monitor_journald': self.journald,
            'scalyr_server': self.scalyr_server,
            'parse_lines_json': self.parse_lines_json,
            'enable_profiling': self.enable_profiling,
        }

        current_paths = self._get_current_log_paths()
        new_paths = {log['path'] for log in self.logs.values()}

        diff_paths = new_paths.symmetric_difference(current_paths)

        if self._first_run or diff_paths:
            logger.debug('Scalyr watcher agent new paths: %s', diff_paths)
            logger.debug('Scalyr watcher agent current paths: %s', current_paths)
            try:
                config = self.tpl.render(**kwargs)

                with open(self.config_path, 'w') as fp:
                    fp.write(config)
            except Exception:
                logger.exception('Scalyr watcher agent failed to write config file.')
            else:
                self._first_run = False
                logger.info('Scalyr watcher agent updated config file %s with %s log targets.',
                            self.config_path, len(diff_paths))

    def _adjust_target_log_path(self, target):
        try:
            src_log_path = target['kwargs'].get('log_file_path')
            application = target['kwargs'].get('application') or target['kwargs'].get('pod_name') or 'none'
            version = target['kwargs'].get('version') or 'none'
            container_id = target['id']

            if not os.path.exists(src_log_path):
                return None

            dst_name = '{}-{}.log'.format(application, version)
            parent = os.path.join(self.dest_path, container_id)
            dst_log_path = os.path.join(parent, dst_name)

            if not os.path.exists(parent):
                os.makedirs(parent)

            # symlink to have our own friendly log file name!
            if not os.path.exists(dst_log_path):
                os.symlink(src_log_path, dst_log_path)

            return dst_log_path
        except Exception:
            logger.exception('Scalyr watcher agent Failed to adjust log path.')
            return None

    def _get_current_log_paths(self) -> list:
        targets = set()

        try:
            if os.path.exists(self.config_path):
                with open(self.config_path) as fp:
                    config = json.load(fp)
                    targets = {log.get('path') for log in config.get('logs', [])}
                    logger.debug('Scalyr watcher agent loaded existing config %s: %d log targets exist!',
                                 self.config_path, len(config.get('logs', [])))
            else:
                logger.warning('Scalyr watcher agent cannot find config file!')
        except Exception:
            logger.exception('Scalyr watcher agent failed to read config!')

        return targets
